crates/bevy_math/src/bezier.rs

```diff
@@ -0,0 +1,253 @@
+    pub fn ease(&self, time: f32) -> f32 {
+        let x = time.clamp(0.0, 1.0);
+        let t = self.find_t_given_x(x);
+        self.evaluate_y_at(t)
+    }
+
+    /// Compute the x-coordinate of the point along the Bezier curve at `t`.
+    #[inline]
+    pub fn evaluate_x_at(&self, t: f32) -> f32 {
+        bezier_impl::position([0.0, self.p1.x, self.p2.x, 1.0], t)
+    }
+
+    /// Compute the y-coordinate of the point along the Bezier curve at `t`.
+    #[inline]
+    pub fn evaluate_y_at(&self, t: f32) -> f32 {
+        bezier_impl::position([0.0, self.p1.y, self.p2.y, 1.0], t)
+    }
+
+    /// Compute the slope of the line at the given parametric value `t` with respect to t.
+    #[inline]
+    pub fn dx_dt(&self, t: f32) -> f32 {
+        bezier_impl::velocity([0.0, self.p1.x, self.p2.x, 1.0], t)
+    }
+
+    /// Solve for the parametric value `t` that corresponds to the given value of `x`.
+    #[inline]
+    pub fn find_t_given_x(&self, x: f32) -> f32 {
+        let mut t_guess = x;
+        (0..Self::MAX_ITERS).any(|_| {
+            let x_guess = self.evaluate_x_at(t_guess);
+            let error = x_guess - x;
+            if error.abs() <= Self::MAX_ERROR {
+                true
+            } else {
+                let slope = self.dx_dt(t_guess);
+                t_guess -= error / slope;
+                false
+            }
+        });
+        t_guess.clamp(0.0, 1.0)
+    }
+}
+
+/// Generic implementations for sampling cubic Bezier curves. Consider using the methods on
+/// [`Bezier`] for more ergonomic use.
+pub mod bezier_impl {
+    use super::Point;
+
+    /// Compute the bernstein basis polynomial for iteration `i`, for a Bezier curve with with
+    /// degree `n`, at `t`.
+    #[inline]
+    pub fn bernstein_basis(n: usize, i: usize, t: f32) -> f32 {
+        (1. - t).powi((n - i) as i32) * t.powi(i as i32)
+    }
+
+    /// Efficiently compute the binomial coefficient
+    #[inline]
+    fn binomial_coeff(n: usize, k: usize) -> usize {
+        let mut i = 0;
+        let mut result = 1;
+        let k = match k > n - k {
+            true => n - k,
+            false => k,
+        };
+        while i < k {
+            result *= n - i;
+            result /= i + 1;
+            i += 1;
+        }
+        result
+    }
+
+    /// Evaluate the Bezier curve B(t) of degree `N-1` at the parametric value `t`.
+    #[inline]
+    pub fn position<P: Point, const N: usize>(control_points: [P; N], t: f32) -> P {
+        let p = control_points;
+        let degree = N - 1;
+        (0..=degree)
+            .map(|i| p[i] * binomial_coeff(degree, i) as f32 * bernstein_basis(degree, i, t))
+            .sum()
+    }
+
+    /// Compute the first derivative B'(t) of Bezier curve B(t) of degree `N-1` at the given
+    /// parametric value `t` with respect to t.
+    #[inline]
+    pub fn velocity<P: Point, const N: usize>(control_points: [P; N], t: f32) -> P {
+        debug_assert!(
+            N > 1,
```

Comment:
Shouldn't this always return 0 for lower dimensional curves instead?

Reply:
Hmmm, that's a good point. We can rely on the `Default` impl returning 0 for all our `Point` types.